### PR TITLE
Write governance items during init process

### PR DESCRIPTION
**Pull request by aidan-kwon**
## Proposed changes

Currently, governance items of the genesis block are stored when the node started.
This change makes the items to be stored during the init process. 
Also, some governance-related code is refactored. 

- `37b4e1d`: Move default config getters to params package
- `a6cd64a`: Refactor initGenesis
- `75f274a`: Write governance items during init process

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [x] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: cmd/utils/nodecmd/chaincmd.go [hunk 1]
```diff
@@ -74,73 +75,51 @@ func initGenesis(ctx *cli.Context) error {
...
+	govItemBytes, err := json.Marshal(govSet.Items())
 	if err != nil {
 		logger.Crit("Failed to json marshaling governance data", "err", err)
 	}
-	if genesis.Governance, err = rlp.EncodeToBytes(gbytes); err != nil {
+	if genesis.Governance, err = rlp.EncodeToBytes(govItemBytes); err != nil {
 		logger.Crit("Failed to encode initial settings. Check your genesis.json", "err", err)
 	}
+
+	params.SetStakingUpdateInterval(genesis.Config.Governance.Reward.StakingUpdateInterval)
```
Comment (KimKyungup) [inline cmd/utils/nodecmd/chaincmd.go:112]: Is it ok not to check if it is 0 or not?

File: params/config.go [hunk 2]
```diff
@@ -20,11 +20,13 @@
 
 package params
 
+import "C"
```
Comment (KimKyungup) [inline params/config.go]: Why was this pkg added?

Comment (aidan-kwon) [reply params/config.go]: I don't know why ... It is removed in `193db0b`

File: cmd/utils/nodecmd/chaincmd.go (hunk 1 above)
Comment (aidan-kwon) [reply cmd/utils/nodecmd/chaincmd.go:112]: If `genesis.Config.Governance.Reward.StakingUpdateInterval` is 0, it will be updated to the default value at `genesis.Config.Enhance()` (line 96)

Comment (aidan-kwon) [reply cmd/utils/nodecmd/chaincmd.go:112]: I also added validation logic for the interval values in `ValidateGenesisConfig` to make sure.

File: cmd/utils/nodecmd/chaincmd.go [hunk 3]
```diff
@@ -74,73 +75,51 @@ func initGenesis(ctx *cli.Context) error {
...
-			}
-		}
-	}
-
-	if genesis.Config.Governance.Reward.StakingUpdateInterval != 0 {
-		params.SetStakingUpdateInterval(genesis.Config.Governance.Reward.StakingUpdateInterval)
-	} else {
-		genesis.Config.Governance.Reward.StakingUpdateInterval = params.StakingUpdateInterval()
+	if genesis.Config == nil {
+		logger.Crit("Invalid genesis config")
```
Comment (winnie-byun) [inline cmd/utils/nodecmd/chaincmd.go:93]:
```suggestion
		logger.Crit("genesis config not set")
```

File: cmd/utils/nodecmd/chaincmd.go [hunk 4]
```diff
@@ -160,117 +139,75 @@ func initGenesis(ctx *cli.Context) error {
...
+
+		logger.Info("Successfully wrote genesis state", "database", name, "hash", hash.String())
+		chainDB.Close()
 	}
 	return nil
 }
 
-func getGovernanceItemsFromGenesis(genesis *blockchain.Genesis) governance.GovernanceSet {
-	g := governance.NewGovernanceSet()
+func ValidateGenesisConfig(g *blockchain.Genesis) error {
```
Comment (winnie-byun) [inline cmd/utils/nodecmd/chaincmd.go:164]: `Enhance` and `ValidateGenesisConfig` do similar thing. How about combining them?

File: cmd/utils/nodecmd/chaincmd.go [hunk 5]
```diff
@@ -74,73 +75,51 @@ func initGenesis(ctx *cli.Context) error {
...
 	if err != nil {
 		logger.Crit("Failed to json marshaling governance data", "err", err)
 	}
-	if genesis.Governance, err = rlp.EncodeToBytes(gbytes); err != nil {
+	if genesis.Governance, err = rlp.EncodeToBytes(govItemBytes); err != nil {
 		logger.Crit("Failed to encode initial settings. Check your genesis.json", "err", err)
 	}
+
+	params.SetStakingUpdateInterval(genesis.Config.Governance.Reward.StakingUpdateInterval)
+	params.SetProposerUpdateInterval(genesis.Config.Governance.Reward.ProposerUpdateInterval)
```
Comment (winnie-byun) [inline cmd/utils/nodecmd/chaincmd.go:113]: Why did you move these code to back?

File: governance/default.go [hunk 6]
```diff
@@ -550,44 +566,8 @@ func (gov *Governance) updateChangeSet(vote GovernanceVote) bool {
 	return false
 }
 
-func GetDefaultGovernanceConfig(engine params.EngineType) *params.GovernanceConfig {
```
Comment (yoomee1313) [inline governance/default.go:553]: Can I ask why you move those getter functions(GetDefault~~) from governance package to params package?

File: governance/default.go [hunk 7]
```diff
@@ -401,6 +402,21 @@ func (gs *GovernanceSet) Merge(change map[string]interface{}) {
...
+		itemCache:                newGovernanceCache(),
+		currentSet:               NewGovernanceSet(),
+		changeSet:                NewGovernanceSet(),
+		lastGovernanceStateBlock: 0,
+		GovernanceTallies:        NewGovernanceTallies(),
+		GovernanceVotes:          NewGovernanceVotes(),
+		idxCacheLock:             new(sync.RWMutex),
+	}
+}
+
```
Comment (winnie-byun) [inline governance/default.go:418]: What is difference between `NewGovernance ` and `NewGovernanceInitialize`?
Could you make comments?

**Review by winnie-byun — CHANGES_REQUESTED** [review]

File: governance/default.go (hunk 6 above)
Comment (aidan-kwon) [reply governance/default.go:553]: `GovernanceConfig` struct is defined in `params` package and `GetDefaultGovernanceConfig` also used variables defined in `params` package, but not in `governance` package. As `GetDefaultGovernanceConfig` is located in `params` package, other packages calling `GetDefaultGovernanceConfig` should import `governance` package as well as `params` package. It increases the complexity of the import chain and may suffer unnecessary package cycle issues later.

File: params/config.go [hunk 8]
```diff
@@ -241,6 +244,49 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, head *big.Int) *Confi
...
+	if c.Clique != nil {
+		if c.Istanbul != nil {
+			return Unknown
+		}
+		return UseClique
+	}
+	if c.Istanbul != nil {
+		return UseIstanbul
+	}
+	return Unknown
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply params/config.go]: Changed. Thanks!

File: params/config.go [hunk 9]
```diff
@@ -338,3 +384,39 @@ func (c *IstanbulConfig) Copy() *IstanbulConfig {
 
 	return newIC
 }
+
+func GetDefaultGovernanceConfig(engine EngineType) *GovernanceConfig {
```
[the remark this answers is not in the document]
Comment (aidan-kwon) [reply params/config.go:387]: I agree with you, but I didn't investigate why the unused parameter was set before. So, I added a comment to consider your suggestion later.

File: cmd/utils/nodecmd/chaincmd.go (hunk 3 above)
Comment (aidan-kwon) [reply cmd/utils/nodecmd/chaincmd.go]: Changed. Thanks!

File: cmd/utils/nodecmd/chaincmd.go (hunk 5 above)
Comment (aidan-kwon) [reply cmd/utils/nodecmd/chaincmd.go:113]: I moved them to increase code readability. With this change, `initGenesis` has a more simple structure: "prepare genesis config" -> "validate config" -> "apply config".

File: governance/default.go (hunk 7 above)
Comment (aidan-kwon) [reply governance/default.go:418]: I added comments.

File: cmd/utils/nodecmd/chaincmd.go (hunk 4 above)
Comment (aidan-kwon) [reply cmd/utils/nodecmd/chaincmd.go:164]: I separated `Enhance` to make it more visual in the code since it forcibly updates users' configuration. It may look a little bit inefficient, however`config.Enhance` is called just once in common cases.
Thanks for the suggestion.

Comment (winnie-byun) [reply cmd/utils/nodecmd/chaincmd.go:164]: How about combining the two functions and add comments instead?
I don't get what your point. Both functions are called once here and seperating the functions will not only make caution but also make the code dirty.

File: governance/default.go [hunk 10]
```diff
@@ -401,6 +402,23 @@ func (gs *GovernanceSet) Merge(change map[string]interface{}) {
...
+// NewGovernanceInitialize creates Governance with the given configuration and read governance state from DB.
+// It also write governance items of the genesis block to DB if any items are stored in DB.
+func NewGovernanceInitialize(chainConfig *params.ChainConfig, dbm database.DBManager) *Governance {
 	ret := Governance{
```
Comment (winnie-byun) [inline governance/default.go:422]:
```suggestion
	governance := *NewGovernance(chainConfig, dbm)
```
How about calling the function inside?

**Comment by aidan-kwon** [comment]
Dear reviewers, I am gonna split this PR into 3 PRs.
